### PR TITLE
GoMod: Respect the replace directive

### PR DESCRIPTION
**Pull request by MarcelBochtler**
Unfortunately this does not fix the issue yet.

ORT uses `go mod vendor -v` to get the vendor modules. For the example project in the FunTest this results in this stderr which is parsed:
```sh
# github.com/fatih/color v1.7.0
github.com/fatih/color
# github.com/mattn/go-colorable v0.1.4
github.com/mattn/go-colorable
# github.com/mattn/go-isatty v0.0.10
github.com/mattn/go-isatty
# golang.org/x/sys v0.0.0-20191008105621-543471e840be
golang.org/x/sys/unix
# golang.org/x/text v0.3.0
golang.org/x/text/internal/tag
golang.org/x/text/language
# rsc.io/quote/v3 v3.1.0 => rsc.io/quote/v3 v3.0.0
rsc.io/quote/v3
# rsc.io/sampler v1.3.0
rsc.io/sampler
```

This PR fixes the parsing of this output and lists `rsc.io/quote/v3 v3.0.0` as vendor module.
This is fixed in this PR and the version version is replaced.

However, to get the dependencies, `go mod graph` is being used:
```sh
❯ go mod graph
github.com/oss-review-toolkit/ort/gomod-synthetic-test-project github.com/fatih/color@v1.7.0
github.com/oss-review-toolkit/ort/gomod-synthetic-test-project github.com/hashicorp/go-multierror@v1.0.0
github.com/oss-review-toolkit/ort/gomod-synthetic-test-project github.com/mattn/go-colorable@v0.1.4
github.com/oss-review-toolkit/ort/gomod-synthetic-test-project github.com/mattn/go-isatty@v0.0.10
github.com/oss-review-toolkit/ort/gomod-synthetic-test-project golang.org/x/oauth2@v0.0.0-20190226205417-e64efc72b421
github.com/oss-review-toolkit/ort/gomod-synthetic-test-project rsc.io/quote/v3@v3.1.0
github.com/hashicorp/go-multierror@v1.0.0 github.com/hashicorp/errwrap@v1.0.0
github.com/mattn/go-colorable@v0.1.4 github.com/mattn/go-isatty@v0.0.8
github.com/mattn/go-isatty@v0.0.10 golang.org/x/sys@v0.0.0-20191008105621-543471e840be
golang.org/x/oauth2@v0.0.0-20190226205417-e64efc72b421 cloud.google.com/go@v0.34.0
golang.org/x/oauth2@v0.0.0-20190226205417-e64efc72b421 golang.org/x/net@v0.0.0-20190108225652-1e06a53dbb7e
golang.org/x/oauth2@v0.0.0-20190226205417-e64efc72b421 golang.org/x/sync@v0.0.0-20181221193216-37e7f081c4d4
golang.org/x/oauth2@v0.0.0-20190226205417-e64efc72b421 google.golang.org/appengine@v1.4.0
rsc.io/quote/v3@v3.1.0 rsc.io/sampler@v1.3.0
github.com/mattn/go-isatty@v0.0.8 golang.org/x/sys@v0.0.0-20190222072716-a9d3bda3a223
google.golang.org/appengine@v1.4.0 github.com/golang/protobuf@v1.2.0
google.golang.org/appengine@v1.4.0 golang.org/x/net@v0.0.0-20180724234803-3673e40ba225
google.golang.org/appengine@v1.4.0 golang.org/x/text@v0.3.0
rsc.io/sampler@v1.3.0 golang.org/x/text@v0.0.0-20170915032832-14c0d48ead0c
```

Here `rsc.io/quote/v3@v3.1.0` is still listed and not replaced by `v3.0.0`.

The way replace directive works is that the source code of the module is replaced, but the version will still be the one added to the `require` directive, see https://github.com/golang/go/issues/40513.

How could ORT handle this? Because IMO it should only list the actually used sources, and not the ones listed in `require`.

### CODE REVIEW

**Comment by sschuberth** [comment]
> How could ORT handle this?

Looks like we need to record the replacements and reapply them to the graph, or?

File: analyzer/src/main/kotlin/managers/GoMod.kt [hunk 1]
```diff
@@ -143,10 +144,21 @@ class GoMod(
...
+                // The package can be replaced using GoMod's replace directive. This should only take the used package
+                // into account, see: https://golang.org/doc/modules/gomod-ref#replace.
+                val replaceIndicator = " => "
+
+                val withoutPrefix = line.withoutPrefix("# ")
+                val moduleVersionLine = if (line.contains(replaceIndicator)) {
+                    withoutPrefix?.substring(line.indexOf(replaceIndicator) + replaceIndicator.length)
+                } else {
+                    withoutPrefix
+                }
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/GoMod.kt:157]: Can we simply do:

                val moduleVersionLine = line.withoutPrefix("# ")?.let {
                    line.split(replaceIndicator).last()
                }

**Review by sschuberth — CHANGES_REQUESTED** [review]

**Comment by MarcelBochtler** [comment]
After some more digging it seems to me that this comment is correct: https://github.com/oss-review-toolkit/ort/issues/3934#issuecomment-852299217

`go mod graph` will not take the actual used dependency into account, while `go list -deps ./...` contains this information:
```
	"Module": {
		"Path": "rsc.io/quote/v3",
		"Version": "v3.1.0",
		"Replace": {
			"Path": "rsc.io/quote/v3",
			"Version": "v3.0.0",
			"Time": "2018-07-09T15:34:46Z",
			"Dir": "/Users/bom1wa3/workspace/go/pkg/mod/rsc.io/quote/v3@v3.0.0",
			"GoMod": "/Users/bom1wa3/workspace/go/pkg/mod/cache/download/rsc.io/quote/v3/@v/v3.0.0.mod"
		},
		"Dir": "/Users/bom1wa3/workspace/go/pkg/mod/rsc.io/quote/v3@v3.0.0",
		"GoMod": "/Users/bom1wa3/workspace/go/pkg/mod/cache/download/rsc.io/quote/v3/@v/v3.0.0.mod"
	},
```

**Comment by MarcelBochtler** [comment]
Superseeded by https://github.com/oss-review-toolkit/ort/pull/5468

**Pull request closed without merging**